### ISSUE TITLE
Add a texture cache for sprites in Video memory 

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags#1869]
**Describe the problem**
When profiling AGS engine, specially in high resolution games, it's noticeable that there's a reasonable amount of time spent in [BitmapToVideoMem](https://github.com/adventuregamestudio/ags/blob/778c898f923ecdad3eb305f306015172df1aeb4d/Engine/gfx/gfxdriverbase.cpp#L397) and [BitmapToVideoMemOpaque](https://github.com/adventuregamestudio/ags/blob/778c898f923ecdad3eb305f306015172df1aeb4d/Engine/gfx/gfxdriverbase.cpp#L562) functions, from the VideoMemoryGraphicsDriver. These functions convert the bitmaps to a format friendly to the graphics driver color depth, with some specific adjustments (e.g., adjustment to texture tiles), and then push these bitmaps to the videocard memory as textures.

As I can understand, right now, when animating something repeatedly on screen, a previous frame that appeared recently must be pushed again, if it's not kept on screen somewhere.

**Suggested change**
Sprite that is loaded directly from the assets doesn't change dynamically in game, it would be nice if the textures created from these could be cached, maybe in a "first level" texture cache.

There may be some caveats

- Video memory varies greatly across videocards, we may need a limit to this cache
- When using dynamic sprites, or other generated sprites (texts?), it should possibly instead use the current method available, ignoring this new cache.
- Maybe if out of the cache memory limit (but required to show a sprite), also devolve to the current available method
- There may be some cases where the video driver requires we empty the cache and recreate all textures (maybe switching from Fullscreen to windowed mode?) Edit: apparently only in Direct3D and with sprites that change render target.
- ~~Caching room backgrounds is probably not very useful~~ Edit: forgot about animated backgrounds! Thanks 
- Not sure if the implementation should allow a case where a sprite is in Texture Cache but not in the spritecache anymore, although we don't need the data, the width and height information may be used elsewhere (need to check), but this mode may be something left for later.

**Additional context**

We already have a short term texture cache, see #1657

**Comment by ivan-mogilko** [comment]
> When using dynamic sprites, or other generated sprites (texts?), it should possibly instead use the current method available, ignoring this new cache.

There's no reason imo why dynamically created sprites should not be a part of this cache. After all, the dynamic sprite may also stay unchanged for a long while in memory, while being assigned, unassigned and re-assigned to various objects.

**Comment by ericoporto** [comment]
Good point on the room animation, I updated above.

Uhm, about dynamic sprites, I mostly didn't remember if we have already some way to know it changed and need to update it's texture. I guess we do since it works currently.

The first implementation of this should probably be very simple, just enough that it can be profiled and experimented with.

**Comment by ivan-mogilko** [comment]
On implementation:

1. Pick out MRU/LRU cache algorithm from SpriteCache in a standalone class. Either make it a base virtual class of a cache, or a template class which lets referring to an abstract object type and is used by the corresponding cache class(es).
2. The texture cache should be storing [`TextureData`](https://github.com/adventuregamestudio/ags/blob/0180e6a3fa8218d3b4d6ede0316a7131ce49b256/Engine/gfx/gfxdriverbase.h#L216-L226) objects. NOTE: DDB is an individual texture "use", many DDBs could share same TextureData and apply different parameters (scale, transparency, and so forth) without modifying texture itself.
3. The size of an object should be calculated by the respective graphics renderer, as TextureData is a base struct, and individual implementations may contain different amount of data. Maybe TextureData struct itself could save a size of memory its members use.
4. The texture id should be sprite id, but we need some idea about things that do not correspond to sprites (yet). Such as room bgs. I see two options: a) reserve certain sprite ids for our purpose, or b) use the highest id bit for a flag that creates custom ids. NOTE: historically normal sprite id is a signed int32, so it's not using full uint32 bits. It cannot be made uint32 in the current circumstances, because then script properties and commands won't be able to refer to them (ags script does not support unsigned int).
5. Add a new config setting for the texture cache limit. Might ask the graphics driver about its actual limit, and clamp the setting.
6. Add a new field to the winsetup for this setting.

**Comment by ivan-mogilko** [comment]
@ericoporto a working experiment, seems stable, but number of things missing:
https://github.com/ivan-mogilko/ags-refactoring/tree/361--texturecache

Was testing Dave's "Old Skies" again. NOTE: I had to hardcode a disabled vsync in the engine to be able to test max fps, because the game sets it on startup and has no way to disable in game menu.
The results are quite curious

| location | 3.6.1.1 | 361--texturecache |
| --- | --- | --- |
| main menu | ~525 | 670-680 |
| bedroom | 550-570 | 670-680 |
| travel room | 590-600 | 670-680 |

To summarize, the texture cache removes something that caused most fps differences between the scenes.

**Comment by ericoporto** [comment]
I noticed in Dave's game significant improvements, specifically the travel room you mention, in my CPU when its usage is low it the fan spin slower and there is audible difference in that room, it went from 9.6-12.8% to 6.4% in the load in the core running AGS - it both reduced and also reduced variation.

In my "Overlay based games" (like Don't Give Up the Cat) I didn't notice changes in performance - neither improvements or issues.

**Comment by ivan-mogilko** [comment]
> In my "Overlay based games" (like Don't Give Up the Cat) I didn't notice changes in performance - neither improvements or issues.

The difference will only be noticeable when the object changes graphic repeatedly. If an object displaying same sprite all the time, then there will be no difference.

**Comment by ericoporto** [comment]
Hey, about the texture cache size, I haven't figured any way to get the available vídeo memory size. I have a feeling that it being configurable similar to how it works with the spritecache is the way to go.

**Comment by ivan-mogilko** [comment]
> Hey, about the texture cache size, I haven't figured any way to get the available video memory size. I have a feeling that it being configurable similar to how it works with the spritecache is the way to go.

Right, there has to be a config setting, but I hoped to find something to know a limit, so that a setting could be clamped to, for safety.
If finding out a memory per texture is complicated or impossible, we could use approximation knowing the texture resolution.

**Comment by ericoporto** [comment]
Sorry, I didn't mean memory for the texture, I meant total free/available memory in the videocard. There used to be this : https://www.libsdl.org/release/SDL-1.2.15/docs/html/sdlvideoinfo.html

But I think this has been removed in SDL2 / 3 and I can't find some crossplatform way when looking online (it appears it even changes between drivers and videocard)
